pkg/wekafs/controllerserver.go

```diff
@@ -327,6 +340,70 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}, nil
 }
 
+// validateFilesystemCapacity checks if requested capacity plus the capacities of existing PVCs would exceed
+// the filesystem total capacity for directory-backed volumes. Returns nil if validation passes.
+func (cs *ControllerServer) validateFilesystemCapacity(ctx context.Context, v *Volume, requestedCapacity int64) error {
+	ctx, span := otel.Tracer(TracerName).Start(ctx, "ValidateFilesystemCapacity")
+	defer span.End()
+
+	logger := log.Ctx(ctx).With().
+		Str("volume_id", v.GetId()).
+		Str("filesystem", v.FilesystemName).
+		Int64("requested_capacity", requestedCapacity).
+		Logger()
+
+	// Only validate directory-backed volumes (dir/v1)
+	// Skip: filesystem-backed, snapshot-backed, and hybrid volumes
+	if !v.hasInnerPath() || v.isOnSnapshot() {
+		logger.Debug().Msg("Skipping capacity validation for non-directory-backed volume")
+		return nil
+	}
+
+	// Get filesystem object for total capacity
+	fsObj, err := v.getFilesystemObj(ctx, true)
+	if err != nil {
+		logger.Error().Err(err).Msg("Failed to fetch filesystem object")
+		return status.Errorf(codes.Internal, "failed to fetch filesystem %s: %v", v.FilesystemName, err)
+	}
+	if fsObj == nil {
+		logger.Error().Msg("Filesystem object is nil")
+		return status.Errorf(codes.Internal, "filesystem %s not found", v.FilesystemName)
+	}
+
+	if cs.pvCache == nil {
+		logger.Warn().Msg("PV cache not available, skipping capacity validation.")
+		return nil
+	}
+
+	// Get existing directory-backed PV capacity from cache
+	sumExistingCapacity := cs.pvCache.GetTotalDirectoryCapacity(ctx, v.FilesystemName)
+
+	logger.Info().
+		Int64("sum_existing_capacity", sumExistingCapacity).
+		Msg("Retrieved existing capacity from PV cache")
+
+	totalRequired := sumExistingCapacity + requestedCapacity
+	logger.Debug().
+		Int64("sum_existing_capacity", sumExistingCapacity).
+		Int64("total_required", totalRequired).
+		Int64("filesystem_capacity", fsObj.TotalCapacity).
+		Msg("Calculated capacity requirements")
+
+	if totalRequired > fsObj.TotalCapacity {
```

Comment:
I am not sure that TotalCapacity is the correct one here.
Might be that there is a better one:
Please check if there is another parameter such as total usable capacity that returns the total possible capacity minus allocated for snapshots.

Not sure, but maybe